Automattic/wp-calypso
# Checkout: Use ResponseCartProduct for WPLineItem

#### Changes proposed in this Pull Request

The `WPCOMCartItem` type was created as a sort of hack to allow passing `ResponseCartItem` object data through checkout via the `useLineItems` hook that's part of the `composite-checkout` package, since the generic product data available in that hook's types is not sufficient. However, now that `useShoppingCart` is available, we can get the `ResponseCartItem` products directly where they are needed. This will make it easier to manipulate and query cart items in the code since there will only be one data type we need to understand, and since `ResponseCartItem` predates `WPCOMCartItem`, there's already considerable support for it in calypso.

This PR refactors the `WPLineItem` (aka: `LineItem`) component in calypso checkout so that it uses a `ResponseCartProduct` instead of a `WPCOMCartItem`. Because this component was also being used for non-product line items like total, subtotal, tax, coupons, and credits, this PR also creates a `NonProductLineItem` component for those.

There's still more work to be done to remove `WPCOMCartItem` entirely, but this will take care of the biggest use of that type: the checkout products list.

In order to be certain that these types were being used correctly, this PR also converts the whole `WPOrderReviewLineItems` file to TypeScript, which resulted in this PR being quite large. It's probably easiest to review the commits one-by-one than all together.

#### Screenshots
Product list:
<img width="493" alt="Screen Shot 2021-02-08 at 5 43 55 PM" src="https://user-images.githubusercontent.com/2036909/107291050-8c06d100-6a35-11eb-9ece-1cfd161e172e.png">

Product variation picker:
<img width="495" alt="Screen Shot 2021-02-08 at 5 44 02 PM" src="https://user-images.githubusercontent.com/2036909/107291058-8f01c180-6a35-11eb-8528-c09021f909a1.png">

Tax, credits, subtotal on last step:
<img width="500" alt="Screen Shot 2021-02-08 at 5 44 35 PM" src="https://user-images.githubusercontent.com/2036909/107291071-945f0c00-6a35-11eb-98ba-6dda8597de48.png">


#### Testing instructions

- Make sure your account has a little bit of money as free credits (not enough to pay for a whole plan).
- Add a plan to your cart and visit checkout.
- Verify that the plan title and price are displayed.
- Verify that the plan's original price is displayed crossed-out. (If you don't have a discount, try to get one somehow to make sure this works.)
- Click "Edit" on the first step and verify that the product variation picker is displayed.
- For the contact form, use a postal code and country that require taxes (eg: 10001, US).
- Verify that the subtotal, total, credits, and taxes are displayed in the last step.
- Add a coupon; verify that it was added successfully and displays with the order.
- Remove the coupon; verify that it was removed successfully.
